lib/pure/os.nim

```diff
@@ -1648,6 +1648,28 @@ proc setFilePermissions*(filename: string, permissions: set[FilePermission],
       var res2 = setFileAttributesA(filename, res)
     if res2 == - 1'i32: raiseOSError(osLastError(), $(filename, permissions))
 
+proc isAdmin*: bool {.noWeirdTarget.} =
+  ## Returns whether the caller's process is a member of the Administrators local
+  ## group (on Windows) or a root (on POSIX), via `geteuid() == 0`.
+  when defined(windows):
+    # Rewrite of the example from Microsoft Docs:
+    # https://docs.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-checktokenmembership#examples
+    var ntAuthority = SID_IDENTIFIER_AUTHORITY(Value: SECURITY_NT_AUTHORITY)
+    var administratorsGroup: PSID
+    var b: WINBOOL = AllocateAndInitializeSid(addr ntAuthority,
+                                              BYTE(2),
+                                              SECURITY_BUILTIN_DOMAIN_RID,
+                                              DOMAIN_ALIAS_RID_ADMINS,
+                                              0, 0, 0, 0, 0, 0,
+                                              addr administratorsGroup)
+    if isSuccess(b):
+      if not isSuccess(CheckTokenMembership(0, administratorsGroup, addr b)):
+        b = 0
+      discard FreeSid(administratorsGroup)
```

Comment:
Consider checking the return value and calling raiseOSError if it fails.

Reply:
1. Which functions you are talking about? I think it's enough to check `AllocateAndInitializeSid` and `CheckTokenMembership`, but not `FreeSid`, because the success of the first two determines the success of `isAdmin`, and the last one is just about free. Also, this is how PostgreSQL (which is a mature source of truth in my opinion) is written: https://doxygen.postgresql.org/win32security_8c.html#ae6b61e106fa5d6c5d077a9d14ee80569
2. Is it OK to raise `raiseOSError(osLastError(), "could not get SID for Administrators group")` and `raiseOSError(osLastError(), "could not check access token membership")`?

Reply:
Done as I've proposed.

Reply:
I mean I don't care a lot about this, but from my side it looks risky to discard this return value, you could end up leaking memory without knowing about it.

Reply:
I looked at other uses of `raiseOSError`, but it's not obvious to me what to pass as arguments to get meaningful error messages.

Reply:
the previous code was not leaking but now can leak:
```nim
    if not isSuccess(CheckTokenMembership(0, administratorsGroup, addr b)):
      raiseOSError(osLastError(), "could not check access token membership")

    if FreeSid(administratorsGroup) != nil:
      raiseOSError(osLastError(), "failed to free SID for Administrators group")
```
`FreeSid(administratorsGroup)` is now not called when `isSuccess(CheckTokenMembership(0, administratorsGroup, addr b)):` fails

there is a way to make it all work